File: L1T1/src/L1T1.py
import numpy as np
import cv2
import oe


def main():
    src = cv2.imread("img/1.png")

    if not src.any():
        print("img is empty")

    kernel = np.ones((3, 3), np.uint8)
    gray = cv2.cvtColor(src, cv2.COLOR_BGR2GRAY)

# black ball proc
    canny = cv2.Canny(gray, 127, 255)
    hole = oe.Hole(canny)
    hole = cv2.erode(hole, kernel, iterations=2)

    # find ipad proc
    contours, _ = cv2.findContours(
        hole, cv2.RETR_LIST, cv2.CHAIN_APPROX_SIMPLE)

    candidate_rect = [rotateRect for rotateRect in
                      [cv2.minAreaRect(contour) for contour in contours]
                      if 50 < abs(rotateRect[2]) < 70]

    if not len(candidate_rect):
        print("no match")

    # find ball proc
    x, y, w, h = oe.boundingRect(candidate_rect[0])
    ipad_roi = canny[y:y+h, x:x+w]
    ipad_roi = oe.Hole(ipad_roi, 1)

    contours, _ = cv2.findContours(
        ipad_roi, cv2.RETR_EXTERNAL, cv2.CHAIN_APPROX_SIMPLE, offset=(x, y))

    candidate_ball = [rotateRect for rotateRect in
                      [cv2.minAreaRect(contour) for contour in contours]
                      if rotateRect[2] == 0]

    if not len(candidate_ball):
        print("no match")

    # sort and draw
    sorted_balls = sorted(candidate_ball, key=lambda ball: ball[0][0])
    for ball in sorted_balls:
        x, y, w, h = oe.boundingRect(ball)
        cv2.rectangle(src, (x, y), (x+w, y+h), (0, 0, 255))

    for idx in range(len(sorted_balls)-1):
        cv2.line(src,
                 (int(sorted_balls[idx][0][0]), int(sorted_balls[idx][0][1])),
                 (int(sorted_balls[idx+1][0][0]),
                  int(sorted_balls[idx+1][0][1])),
                 (0, 0, 255))

# map proc
    _, binary = cv2.threshold(gray, 127, 255, cv2.THRESH_BINARY)
    binary = cv2.erode(binary, kernel)

    # find map proc
    contours, _ = cv2.findContours(
        binary, cv2.RETR_LIST, cv2.CHAIN_APPROX_SIMPLE)

    maxArea, maxArea_idx = 0, 0
    for idx in range(len(contours)):
        area = cv2.contourArea(contours[idx])
        if maxArea < area:
            maxArea = area
            maxArea_idx = idx

    x, y, w, h = cv2.boundingRect(contours[maxArea_idx])
    map_roi = gray[y:y+h, x:x+w]

    # show and draw
    map_roi = cv2.Canny(map_roi, 0, 2)
    map_roi = cv2.dilate(map_roi, kernel)

    mask = np.zeros((src.shape[0]+2, src.shape[1]+2), np.uint8)
    mask[y:y+h, x:x+w] = map_roi
    cv2.floodFill(src, mask, (x, y), (255, 255, 255))

# swap computer monitor
    canny = cv2.Canny(gray, 127, 255)
    canny = cv2.dilate(canny, kernel)

    contours, hierarchy = cv2.findContours(
        canny, cv2.RETR_TREE, cv2.CHAIN_APPROX_NONE)
    for i in range(len(contours)):
        for j in range(i + 1, len(contours)):
            if hierarchy[0][i][2] == -1 or hierarchy[0][i][3] == -1:
                break

            contour_i = cv2.boundingRect(contours[i])
            contour_j = cv2.boundingRect(contours[j])

            if contour_i[2] < 100 or contour_i[3] < 100:
                break

            if contour_j[2] < 100 or contour_j[3] < 100:
                continue

            if cv2.pointPolygonTest(contours[i], (contour_j[0]+(contour_j[2]//2), contour_j[1]+(contour_j[3]//2)), False) != -1:
                continue

            if not oe.approxLenth((contour_i[2], contour_j[2]), 0.1):
                continue

            if not oe.approxLenth((contour_i[3], contour_j[3]), 0.1):
                continue

            source = cv2.matchShapes(
                contours[i], contours[j], cv2.CONTOURS_MATCH_I1, 0)

            if source == 0 or source > 1:
                continue

            left, right = (contour_i, contour_j) if contour_i[0] < contour_j[0] else (contour_j, contour_i)

    left_monitor = src[left[1]:left[1]+left[3], left[0]:left[0]+left[2]]
    right_monitor = src[right[1]:right[1] +
                        right[3], right[0]:right[0]+right[2]].copy()

    src[right[1]:right[1]+right[3], right[0]:right[0]+right[2]] = left_monitor
    src[left[1]:left[1]+left[3], left[0]:left[0]+left[2]] = right_monitor

    cv2.rectangle(
        src, (left[0], left[1]), (left[0]+left[2], left[1]+left[3]), (0, 255, 0))
    cv2.rectangle(
        src, (right[0], right[1]), (right[0]+right[2], right[1]+right[3]), (255, 0, 0))

    cv2.imshow("src", src)
    cv2.imshow("map_roi", map_roi)
    cv2.imshow("ipad_roi", ipad_roi)

    while True:
        if cv2.waitKey(1) == 27:
            break

    cv2.destroyAllWindows()


if __name__ == '__main__':
    main()

File: L1T1/src/L1T1.cpp
#include <iostream>
#include <vector>

#include <opencv4/opencv2/opencv.hpp>
#include <opencv4/opencv2/highgui/highgui.hpp>
#include <opencv4/opencv2/imgproc/imgproc.hpp>
#include <opencv4/opencv2/core/core.hpp>

bool RectXCmp(const cv::RotatedRect &first, const cv::RotatedRect &second)
{
    if (first.center.x < second.center.x)
        return true;
    return false;
}

void Hole(const cv::Mat src, cv::Mat &dst, int mode = 0)
{
    cv::Size size = src.size();
    cv::Mat cut, temp;

    temp = cv::Mat::zeros(size.height + 2, size.width + 2, src.type());
    src.copyTo(temp(cv::Range(1, size.height + 1), cv::Range(1, size.width + 1)));
    cv::floodFill(temp, cv::Point(0, 0), cv::Scalar(255));
    temp(cv::Range(1, size.height + 1), cv::Range(1, size.width + 1)).copyTo(cut);

    if (mode == 0)
        dst = src | (~cut);
    else if (mode == 1)
        dst = ~cut;
}

int main(int argc, char const *argv[])
{
    cv::Mat src = cv::imread("img/1.png");

    if (src.empty())
    {
        std::cout << "img is empty" << std::endl;
        return -1;
    }

    cv::Mat kernel = cv::getStructuringElement(CV_SHAPE_RECT, cv::Size(3, 3));

    cv::Mat gray;
    cv::cvtColor(src, gray, CV_BGR2GRAY);

// black ball process
    cv::Mat ballcanny, hole;
    cv::Canny(gray, ballcanny, 127, 255);
    Hole(ballcanny, hole);
    cv::erode(hole, hole, kernel, cv::Point(-1, -1), 2);

    // find ipad proc
    std::vector<std::vector<cv::Point>> contours;
    cv::findContours(hole, contours, CV_RETR_LIST, CV_CHAIN_APPROX_SIMPLE);

    std::vector<cv::Rect> candidate_rect;
    cv::RotatedRect rotate;
    for (size_t idx = 0; idx < contours.size(); idx++)
    {
        rotate = cv::minAreaRect(contours[idx]);
        double angle = abs(rotate.angle);

        if (50 < angle && angle < 70)
            candidate_rect.push_back(rotate.boundingRect());
    }

    if (candidate_rect.size() < 1)
    {
        std::cout << "no match" << std::endl;
        return -1;
    }

    // find ball proc
    cv::Mat ipad_roi = ballcanny(candidate_rect[0]);
    Hole(ipad_roi, ipad_roi, 1);

    cv::findContours(ipad_roi, contours, CV_RETR_EXTERNAL, CV_CHAIN_APPROX_SIMPLE,
                     cv::Point(candidate_rect[0].x, candidate_rect[0].y));

    std::vector<cv::RotatedRect> candidate_ball;
    for (size_t idx = 0; idx < contours.size(); idx++)
    {
        rotate = cv::minAreaRect(contours[idx]);
        double angle = abs(rotate.angle);

        if (angle == 0)
        {
            cv::drawContours(src, contours, idx, cv::Scalar(0, 0, 255));
            candidate_ball.push_back(rotate);
        }
    }

    // sort and draw
    std::sort(candidate_ball.begin(), candidate_ball.end(), RectXCmp);
    for (size_t idx = 0; idx < candidate_ball.size() - 1; idx++)
    {
        // std::cout << " x[" << idx << "]: " << candidate_ball[idx].x << std::endl;
        cv::line(src,
                 cv::Point(candidate_ball[idx].center.x, candidate_ball[idx].center.y),
                 cv::Point(candidate_ball[idx + 1].center.x, candidate_ball[idx + 1].center.y), 
                 cv::Scalar(0, 0, 255));
    }

// find map process

    cv::Mat bin;
    cv::threshold(gray, bin, 127, 255, cv::THRESH_BINARY);
    cv::erode(bin, bin, kernel);

    // find map proc
    contours.clear();
    cv::findContours(bin, contours, CV_RETR_LIST, CV_CHAIN_APPROX_SIMPLE);
    double maxContour = 0;
    int maxContour_idx = 0;
    for (size_t i = 0; i < contours.size(); i++)
    {
        double area = cv::contourArea(contours[i]);
        if (maxContour < area)
        {
            maxContour = area;
            maxContour_idx = i;
        }
    }
    cv::Rect maxContour_rect = cv::boundingRect(contours[maxContour_idx]);
    cv::Mat map_roi = gray(maxContour_rect);

    // show and draw
    cv::Mat map_canny;
    cv::Canny(map_roi, map_canny, 0, 2);
    cv::dilate(map_canny, map_canny, kernel);

    cv::Mat zeros = cv::Mat::zeros(src.rows + 2, src.cols + 2, CV_8UC1);
    int x = maxContour_rect.x;
    int y = maxContour_rect.y;
    map_canny.copyTo(zeros(maxContour_rect));
    cv::floodFill(src, zeros, cv::Point(x, y), cv::Scalar(255, 255, 255));

    while (true)
    {
        cv::imshow("src", src);
        cv::imshow("map_canny", map_canny);
        cv::imshow("ipad_roi", ipad_roi);

        if (cv::waitKey(1) == 27)
            break;
    }

    cv::destroyAllWindows();
    return 0;
}
